File: src/dotnet/BlockDirectedAcyclicGraph/BlockIndex.cs
using Gma.DataStructures;
using System;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Linq;

namespace NickStrupat
{
	// no verification of block signatures nor chaining is done by the index
	public sealed class BlockIndex
	{
		private readonly Dictionary<Signature, Block> blocksBySignature = new Dictionary<Signature, Block>();
		private readonly Dictionary<Signature, OrderedSet<Block>> childBlocksByParentSignature = new Dictionary<Signature, OrderedSet<Block>>();

		public Int32 Count => blocksBySignature.Count;

		public Boolean Contains(Signature signature) =>
			blocksBySignature.ContainsKey(signature);

		public Boolean TryGetBySignature(Signature signature, out Block block) =>
			blocksBySignature.TryGetValue(signature, out block);

		public Boolean TryGetChildren(Signature signature, [NotNullWhen(true)] out IReadOnlyList<Block>? children)
		{
			if (childBlocksByParentSignature.TryGetValue(signature, out var values))
			{
				children = values.ToList();
				return true;
			}
			children = null;
			return false;
		}

		public void Add(Block block)
		{
			try
			{
				blocksBySignature.Add(block.Signature, block);

				if (block.ParentSignatures.IsEmpty)
					AddOrThrow(default, block);
				else
					foreach (var parentSignature in block.ParentSignatures.AsImmutableSpan())
						AddOrThrow(parentSignature, block);

				void AddOrThrow(Signature parentSignature, Block block)
				{
					if (!childBlocksByParentSignature.TryAddValue(parentSignature, block, Block.EqualityComparer.Default))
						throw new ArgumentException("An element with the same key already exists in the set");
				}
			}
			catch
			{
				blocksBySignature.Remove(block.Signature);
				if (block.ParentSignatures.IsEmpty)
					RemoveIfExists(default, block);
				else
					foreach (var parentSignature in block.ParentSignatures.AsImmutableSpan())
						RemoveIfExists(parentSignature, block);
				throw;

				void RemoveIfExists(Signature parentSignature, Block block)
				{
					if (childBlocksByParentSignature.TryGetValue(parentSignature, out var set))
						if (set.Remove(block) && set.Count == 0)
							childBlocksByParentSignature.Remove(parentSignature);
				}
			}
		}

		public IReadOnlyList<Block> GetAllBlocks() =>
			blocksBySignature.Values.ToList();

		public IEnumerable<(Block Block, UInt32 Level)> TraverseDepthFirst(Block root) => root.TraverseDepthFirst(
			x => TryGetChildren(x.Signature, out var children) ? children : Array.Empty<Block>()
		);

		public IEnumerable<(Block Block, UInt32 Level)> TraverseBreadthFirst(Block root) => root.TraverseBreadthFirst(
			x => TryGetChildren(x.Signature, out var children) ? children : Array.Empty<Block>()
		);
	}
}

File: src/typescript/BlockDirectedAcyclicGraph/BlockIndex.ts
import { Block } from "./Block";
import { Convert } from "../Common/Convert";
import { NodeLevel, Traverse } from "../Common/Traverse";
import { Buffer } from "buffer";

function tryAddValue(map: Map<string, Map<string, Block>>, key: string, block: Block): boolean {
	const values = map.get(key);
	const key2 = Convert.toBase64String(block.parentSignatures)
	if (values !== undefined) {
		if (values.has(key2))
			return false;
		values.set(key2, block);
		return true;
	}
	const map2 = new Map<string, Block>();
	map2.set(key2, block);
	map.set(key, map2);
	return true;
}

export class BlockIndex {
	private readonly blocksBySignature = new Map<string, Block>();
	private readonly childBlocksByParentSignature = new Map<string, Map<string, Block>>();

	contains(signature: Uint8Array): boolean {
		const key = Convert.toBase64String(signature);
		return this.blocksBySignature.get(key) !== undefined;
	}

	tryGetBySignature(signature: Uint8Array): Block | undefined {
		const key = Convert.toBase64String(signature);
		return this.blocksBySignature.get(key);
	}

	tryGetChildren(signature: Uint8Array): readonly Block[] | undefined {
		const key = Convert.toBase64String(signature);
		const set = this.childBlocksByParentSignature.get(key);
		if (set === undefined)
			return undefined;
		return Array.from(set.values());
	}

	add(block: Block): void {
		const key = Convert.toBase64String(block.signature);
		try {
			if (this.blocksBySignature.has(key))
				throw new Error("An element with the same key already exists in the set");
			this.blocksBySignature.set(key, block);

			const addOrThrow = (parentSignature: Buffer, block: Block): void => {
				const psk = Convert.toBase64String(parentSignature);
				if (!tryAddValue(this.childBlocksByParentSignature, psk, block))
					throw new Error("An element with the same key already exists in the set");
			};

			if (block.parentSignatures.length == 0)
				addOrThrow(block.parentSignatures, block);
			else
				for (let parentSignature of block.parentSignaturesIterable())
					addOrThrow(parentSignature, block);
		}
		catch (err) {
			this.blocksBySignature.delete(key);

			const reomveIfExists = (parentSignature: Buffer, block: Block): void => {
				const psk = Convert.toBase64String(parentSignature);
				const childBlocks = this.childBlocksByParentSignature.get(psk);
				if (childBlocks !== undefined) {
					const key2 = Convert.toBase64String(block.parentSignatures)
					if (childBlocks.delete(key2) && childBlocks.size === 0)
						this.childBlocksByParentSignature.delete(psk)
				}
			};

			if (block.parentSignatures.length == 0)
				reomveIfExists(block.parentSignatures, block);
			else
				for (let parentSignature of block.parentSignaturesIterable())
					reomveIfExists(parentSignature, block);

			throw err;
		}
	}

	getAllBlocks(): readonly Block[] {
		return Array.from(this.blocksBySignature.values());
	}

	private getChildren(node: Block): IterableIterator<Block> {
		const children = this.tryGetChildren(node.signature);
		if (children === undefined)
			return [].values();
		return children.values();
	}

	traverseDepthFirst(root: Block): IterableIterator<NodeLevel<Block>> {
		return Traverse.depthFirst(root, this.getChildren);
	}

	traverseBreadthFirst(root: Block): IterableIterator<NodeLevel<Block>> {
		return Traverse.breadthFirst(root, this.getChildren);
	}
}